cdap-app-fabric/src/main/java/co/cask/cdap/internal/app/runtime/workflow/CustomActionExecutor.java

```diff
@@ -0,0 +1,183 @@
+  private final ProgramRunId workflowRunId;
+  private final BasicWorkflowContext workflowContext;
+  private final WorkflowAction action;
+
+  /**
+   * Creates instance which will be used to initialize, run, and destroy the custom action.
+   * @param workflowRunId the Workflow run which started the execution of this custom action.
+   * @param workflowContext an instance of context
+   * @param instantiator to instantiates the custom action class
+   * @param classLoader used to load the custom action class
+   * @throws Exception when failed to instantiate the custom action
+   */
+  public CustomActionExecutor(ProgramRunId workflowRunId, BasicWorkflowContext workflowContext,
+                              InstantiatorFactory instantiator, ClassLoader classLoader) throws Exception {
+    this.workflowRunId = workflowRunId;
+    this.workflowContext = workflowContext;
+    this.action = createAction(workflowContext, instantiator, classLoader);
+  }
+
+  @SuppressWarnings("unchecked")
+  private WorkflowAction createAction(BasicWorkflowContext context, InstantiatorFactory instantiator,
+                                      ClassLoader classLoader) throws Exception {
+    Class<?> clz = Class.forName(context.getSpecification().getClassName(), true, classLoader);
+    Preconditions.checkArgument(WorkflowAction.class.isAssignableFrom(clz), "%s is not a WorkflowAction.", clz);
+    WorkflowAction action = instantiator.get(TypeToken.of((Class<? extends WorkflowAction>) clz)).create();
+    Reflections.visit(action, action.getClass(),
+                      new PropertyFieldSetter(context.getSpecification().getProperties()),
+                      new DataSetFieldSetter(context),
+                      new MetricsFieldSetter(context.getMetrics()));
+    return action;
+  }
+
+  void execute() throws Exception {
+    CountDownLatch executorTerminateLatch = new CountDownLatch(1);
+    ExecutorService executorService = createExecutor(1, executorTerminateLatch, "action-"
+      + workflowContext.getSpecification().getName() + "-%d");
+
+    try {
+      // Run the action in new thread
+      Future<?> future = executorService.submit(new Callable<Void>() {
+        @Override
+        public Void call() throws Exception {
+          setContextCombinedClassLoader(action);
+          try {
+            initializeInTransaction(action, workflowContext);
+            runInTransaction(action, workflowContext);
+            return null;
+          } finally {
+            destroyInTransaction(action, workflowContext);
+          }
+        }
+      });
+      future.get();
+    } catch (Throwable t) {
+      LOG.error("Error executing the action {} in the Workflow {}.", workflowContext.getSpecification().getName(),
+                workflowRunId, t);
+      Throwables.propagateIfPossible(t, Exception.class);
+      throw Throwables.propagate(t);
+    } finally {
+      executorService.shutdownNow();
+      executorTerminateLatch.await();
```

Comment:
Also, why need to use an executor in this method? You create an executor, submit one task and then block on it. It's basically the same as just run whatever is inside the Callable.